### PR TITLE
New build system

### PR DESCRIPTION
**Pull request by janstary**
This gets rid of GNU autotools and introduces a simple, hand-written `./configure`
that uses a set of trivial `have-*.c` programs to test whether e.g. `err()` exists,
or whether `socket()` needs `-lsocket`, etc. See the README.

### CODE REVIEW

**Comment by janstary** [comment]
Merged `master`, now there are no conflicts.

This builds for me on OpenBSD, FeeBSD, Debian, MacOS 10.13.3, MacOS 10.5.8.
I still have the Solaris `hsearch` to do and a bunch of other things, but it basically works.

**Comment by janstary** [comment]
(Sorry for the multiple merge, I firstly merged _my_ behind master before merging the _upstream_ master.)

**Comment by janstary** [comment]
Feel free to squash this into one huge diff, which is very bad practice of course :-)

**Comment by janstary** [comment]
Marged again, to catch up with the *.suo ignore

**Comment by njh** [comment]
Removing autoconf/automake seems like a backwards step.

Why re-invent the wheel, when someone has already done all the hard work?

**Comment by janstary** [comment]
On the contrary, removing the GNU autotools is a forward step out of the nineties.
Not to mention we drop the last GPL dependency and are completely BSD licensed.

**Comment by njh** [comment]
For a project that is still trying to ensure that SunOS builds work, I am not sure saying that autotools is from the nineties is the best justification for getting rid of a very mature build system.

But it is your project, so I will stay quiet now.

**Comment by janstary** [comment]
SunOS sure contains some old cruft. That has nothing to do with RTP tools itself.
If we want to build on SunOS (which we do), the way to deal with it is not to bring in
_more_ ancient cruft.

For example, this is an obvious way to check whether the system has `socket(2)`:

```
#include <sys/socket.h>

int
main(void)
{
        int sock;
        if (-1 == (sock = socket(AF_INET, SOCK_DGRAM, 0)))
                return 1;
        return 0;
}
```

If this code compiles, runs, and returns zero, you have a `socket()` function.
If it doesn't, you don't. (In which case try again with `-lsocket`, because Solaris).

Have you looked at how the GNU autotools actually check for that? Don't.

**Comment by janstary** [comment]
For your reading pleasure:

```
{
  as_lineno=${as_lineno-"$1"} as_lineno_stack=as_lineno_stack=$as_lineno_stack
  { $as_echo "$as_me:${as_lineno-$LINENO}: checking for $2" >&5
$as_echo_n "checking for $2... " >&6; }
if eval \${$3+:} false; then :
  $as_echo_n "(cached) " >&6
else
  cat confdefs.h - <<_ACEOF >conftest.$ac_ext
/* end confdefs.h.  */
/* Define $2 to an innocuous variant, in case <limits.h> declares $2.
   For example, HP-UX 11i <limits.h> declares gettimeofday.  */
#define $2 innocuous_$2

/* System header to define __stub macros and hopefully few prototypes,
    which can conflict with char $2 (); below.
    Prefer <limits.h> to <assert.h> if __STDC__ is defined, since
    <limits.h> exists even on freestanding compilers.  */

#ifdef __STDC__
# include <limits.h>
#else
# include <assert.h>
#endif

#undef $2

/* Override any GCC internal prototype to avoid an error.
   Use char because int might match the return type of a GCC
   builtin and then its argument prototype would still apply.  */
#ifdef __cplusplus
extern "C"
#endif
char $2 ();
/* The GNU C library defines this for functions which it implements
    to always fail with ENOSYS.  Some functions are actually named
    something starting with __ and the normal name is an alias.  */
#if defined __stub_$2 || defined __stub___$2
choke me
#endif

int
main ()
{
return $2 ();
  ;
  return 0;
}
_ACEOF
if ac_fn_c_try_link "$LINENO"; then :
  eval "$3=yes"
else
  eval "$3=no"
fi
rm -f core conftest.err conftest.$ac_objext \
    conftest$ac_exeext conftest.$ac_ext
fi
eval ac_res=\$$3
               { $as_echo "$as_me:${as_lineno-$LINENO}: result: $ac_res" >&5
$as_echo "$ac_res" >&6; }
  eval $as_lineno_stack; ${as_lineno_stack:+:} unset as_lineno

}
```

That's how autotools check ehether a given function exists.
Do you seriously find that better in any sense?

**Comment by janakj** [comment]
The snippet from autotools is actually not all that different from your code. A major problem of both is that you have to compile and **run** a program to determine whether a particular library function is available. This can get very tricky, especially if you need to cross-compile the program, e.g., to build arm binaries on an intel box. A lot of what you call "cruft" originates from the need to support various cross-compilation scenarios in autotools.

Having said that, I do see a point in moving away from autotools. Rtptools is a small and easy to build project and probably does not benefit from using autotools that much anyway.

**Comment by janstary** [comment]
You are right about the cross-compilation, I never took that into account.
But I believe autotools have the very same problem. Do we want/need to
cross-compile somewhere?

Moving away from autotools to keep thing small and simple is the whole point.
RTP tools is a neat little application that consists of ~4600 lines of actual C,
but uses > 7000 lines of arcane shell code to even build.

**Comment by janstary** [comment]
This is the shell code that runs the test in the proposed `./configure`:

```
singletest() {
        cat 1>&3 << __HEREDOC__
${1}${3}: testing...
${COMP} -o have-${1} have-${1}.c ${3}
__HEREDOC__

        if ${COMP} -o "have-${1}" "${SOURCEDIR}/have-${1}.c" ${3} 1>&3 2>&3
        then
                echo "${1}${3}: ${CC} succeeded" 1>&3
        else
                echo "${1}${3}: ${CC} failed with $?" 1>&3
                echo 1>&3
                return 1
        fi

        if ./have-${1} 1>&3 2>&3; then
                echo "${1}${3}: yes" 1>&2
                echo "${1}${3}: yes" 1>&3
                echo 1>&3
                eval HAVE_${2}=1
                rm "have-${1}"
                return 0
        else
                echo "${1}${3}: execution failed with $?" 1>&3
                echo 1>&3
                rm "have-${1}"
                return 1
        fi
}
```

**Comment by mcd500** [comment]
There are several topics need to be discussed for this pull request.

Having a script for generating Makefile for different system will make everybody easier to run the rtptools on different platform. 
The rtptools is used as the reference code for who would like to start write program based on RFC3650 and rtptools historically focusing to be on multiple platform was strong value for the purpose.

The release 1.22 is using autotols, just because I wanted to minimize the breakage and I was trying to be conservative when fixing the build adopting to circumstance from the older releases.
In the 1.88 era, most of them were 32bit systems. I personally do not have any strong preference what  build system to use for it.

When @janstary was referring the new build system, I thought he was going to use cmake, waf  or etc which already exists. Then I saw it was brand new configure script written from scratch. :)
(We may end of making a new build system spawned from rtptools project)

As @janakj mentioned, considering the size of rtptools, the auotools is overkill.
However, I strongly prefer to have cross-compositing capability.
I just checked it after removing the clang detection in configure.ac the following works fine on 1.22 release.

$ ./configure --host=arm-linux-gnueabihf
$ make clean
$ make
$ file rtptrans
rtptrans: ELF 32-bit LSB executable, ARM, EABI5 version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux-armhf.so.3, for GNU/Linux 2.6.32, BuildID[sha1]=cbf96a80fd1bf52487f6a5ef1d33d177142beaf6, not stripped

With
$ ./configure --host=aarch64-linux-gnu
then
$ file rtptransrtptrans: ELF 64-bit LSB executable, ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux-aarch64.so.1, for GNU/Linux 3.7.0, BuildID[sha1]=016631761b0156d6ac548e8dad629755b8b05e83, not stripped

It looks like just changing the compiler  in the CC which might not be so difficult to adapt on new configure script.

Other topics here is that license of autotools which is GPL. 
The post here may related to the autotool.
https://github.com/columbia-irt/rtptools/issues/51

There is no problem in point of license that autotools is being GPL for OpenBSD.
Using GPL programs do not effect of what license to be used for generated programs.
For example, the source which is not GPL and using gcc to compile and the binary would not be GPL is perfectly fine. The GPL do not enforce other program to be GPLunless it contains some portion of GPL sources. Also the none GPL binary being executing while there are GPL binaries but having the context switch, e.g, in other process, between kernal and userlan, also do not have to be GPL. 

The autotools generates `configure` from `configure.am` and `Makefile.am`.
And the generated `configure` will generate `Makefile`.
At this point `configure` uses `aclocal.m4`, `missing`, `install-sh`, `config.sub`, `config.guess`, `compile, `depcomp, which are all GPL programs.
However, all `configure`, `configure.am`, `Makefile.am`, `Makefile` are not GPL and do not have to be GPL from legal point of view of the GPL license.

I understand that FreeBSD and OpenBSD project are going away from GPL programs, and replacing glibc, gcc and others. 
Then they may not like having GPL programs, such as, `aclocal.m4`, `missing`, `install-sh`, `config.sub`, `config.guess`, `compile, `depcomp’, in the tar file of rtptools.
Then we have to remove `aclocal.m4`, `missing`, `install-sh`, `config.sub`, `config.guess`, `compile, `depcomp’  from the rtptools package.

**Comment by janstary** [comment]
The build system should be a thin layer upon the actual code.
What we have now is the exact opposite, hence this proposal.

The build system introduced here is not new and is not my invention;
it comes directly from http://mandoc.bsd.lv/ (the extremely portable
manpage formatter in *BSD, MacOS, and a couple of Linux distros).
It is not my whim, it is well setablished.

Regarding cross-compiling: it is the compiler's job to produce code for the given platform.
The build system's (i.e. ./configure's) role in this is to pass the option. I can add that,
but I don't think it's worth it at all: do we need to cross compile? Do we have a platform
on which we would like to run, but don't have a native compiler for? Currently, we provide
the source tarball (of course), a Windows binary (Intel only), and a RPM (Intel only).
I don't think cross-compiling is an argument against this.

Regarding license: what I mean by acceptable license in https://github.com/columbia-irt/rtptools/issues/51
is the change from the Columbia license (1.21) to BSD (1.22).
GPL itself is not a problem, but of course BSD is preferable.

In autotools (as I understand it), autoconf produces `./configure` from `./configure.ac`,
automake produces `Makefile.in` from `Makefile.am`,
and then `./configure` produces `Makefile` from `Makefile.in`, with the help of 
`aclocal.m4`, `missing`, `install-sh`, `config.sub`, `config.guess`, `compile`, `depcomp`.

I find that quite overingeneered. In the proposed system, you have a static `Makefile`
and a static `./configure` script which produces `Makefile.local` and `config.h`
with the help of `have-*.c`.

**Comment by janakj** [comment]
Regarding cross-compilation: I don't know if you need it. It was my impression that you care about supporting multiple operating systems and platforms. If that's the case, providing support for cross-compilation would make sense. It's not that difficult to do in a small project like rtptools, one just needs to keep in mind that programs may need to be compiled and run in the configuration phase.

In my opinion, having cross-compilation support is useful, because it helps build the software for various embedded platforms, e.g., those based ARM SoCs or Android. I hear those are popular these days.

**Comment by janstary** [comment]
Supporting multiple operating systems and platforms is definitely the goal,
but that's not the same as cross-compiling. Do we know of a OS/platform
where we would like to run, but can't compile natively (and need to cross-compile)?
Running RTP tools on e.g. Android seems a bit far-fetched.

**Comment by janstary** [comment]
So what happens now? Do we want this?
Do we want to stay with GNU autotools?
Who's gonna make that decision?

File: README.md [hunk 1]
```diff
@@ -21,24 +21,63 @@ Refer to the individual manpages for details.
...
+Read the standard output and `Makefile.local`.
+If these look different from what you expected,
+read `configure.local.example`, create `configure.local`,
+and run `./configure` again.
+
+The build system does not use GNU autoconf. All modern operating systems
+are now reasonably close to POSIX and do not need arcane shell magic
+any longer. If your system does need such magic, consider upgrading
+to reasonably modern POSIX-compliant tools rather than asking for
+autoconf-style workarounds.
```
Comment (janakj) [inline README.md]: You don't like autotools and that's fine. But is this text really necessary? README.md is one of the first files a new user encounters.

File: Makefile.depend [hunk 2]
```diff
@@ -0,0 +1,10 @@
+host2ip.o: host2ip.c sysdep.h
+hpt.o: hpt.c sysdep.h
+hsearch.o: hsearch.c sysdep.h hsearch.h
+multimer.o: multimer.c multimer.h notify.h sysdep.h
+notify.o: notify.c sysdep.h notify.h multimer.h ansi.h
+rd.o: rd.c rtpdump.h sysdep.h
+rtpdump.o: rtpdump.c rtp.h sysdep.h vat.h rtpdump.h ansi.h
+rtpplay.o: rtpplay.c sysdep.h notify.h rtp.h rtpdump.h multimer.h ansi.h
+rtpsend.o: rtpsend.c notify.h rtp.h sysdep.h multimer.h ansi.h
+rtptrans.o: rtptrans.c rtp.h sysdep.h rtpdump.h ansi.h notify.h multimer.h vat.h
```
Comment (janakj) [inline Makefile.depend:10]: Does this mean that the list of compile-time dependencies is fixed? Are developers expected to keep this file synchronized with the source code?

Wouldn't it be better and more robust to let the compiler (gcc) generate the list of dependencies automatically?

File: README.md [hunk 3]
```diff
@@ -21,24 +21,63 @@ Refer to the individual manpages for details.
...
+to test the libraries installed on your system, and the standard
+output and standard error output these commands produce.
+Failures are most likely to happen if headers or libraries
+are installed in unusual places or interfaces defined
+in unusual headers. You can also look at `config.h` and
+check that the `#define HAVE_*` lines match your expectations.
+
+### build
+
+Once the source is configured as above, run `make` to build SoX.
```
Comment (janakj) [inline README.md]: SoX -> rtptools?

File: have-gethostbyname.c [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include <netdb.h>
+
+int
+main(void)
+{
+	struct hostent* host;
+	host = gethostbyname("columbia.edu");
```
Comment (janakj) [inline have-gethostbyname.c]: Is it a good idea to invoke a function that internally performs a DNS request? Would it not be possible to test for the presence of gethostbyname without invoking it (and waiting for the DNS response to arrive or timeout)?

If gethostbyname needs to be invoked, perhaps the test could pass an IPv4 address to it, e.g., "127.0.0.1", or a name that is unlikely to trigger name resolution such as "localhost".

File: README.md (hunk 1 above)
Comment (janstary) [reply README.md]: I have no objections to removing this. I felt it appropriate to inform the user
that the `./configure` script is not what he might be expecting from elsewhere.
I will remove it.

File: Makefile.depend (hunk 2 above)
Comment (janstary) [reply Makefile.depend:10]: This list is generated by `mkdep`. It has not changed in a decade,
and can be redone with `make depend`

File: README.md (hunk 3 above)
Comment (janstary) [reply README.md]: copypasta - fixed, thanks.

File: have-gethostbyname.c (hunk 4 above)
Comment (janstary) [reply have-gethostbyname.c]: Yes, `localhost` is probably better. Thanks.

File: Makefile.depend (hunk 2 above)
Comment (janakj) [reply Makefile.depend:10]: I see. I think make depend is fine (as long as there is one command to update the dependency list).

**Comment by mcd500** [comment]
> Regarding license: what I mean by acceptable license in #51
> is the change from the Columbia license (1.21) to BSD (1.22).
> GPL itself is not a problem, but of course BSD is preferable.

Thanks for clarification of your intention regarding the license and the
https://github.com/columbia-irt/rtptools/issues/51

I think we could close of having consensus on the license topics.

**Comment by mcd500** [comment]
For the cross-compilation, almost nobody likes cross-compilation for the developing products.
The most of the reasons of using cross-compilation are;
(*) The target cpu is so slow for compiling and building on PC will dramatically reduce the time of CI loop. Most of the time for product development, time is money, The PC on Intel/AMD is the fastest machine to build at the moment on most cases.
(*) The board with target cpu is not able to even build by variety of limitations, often have too small memory size which result having OOMKiller on the target board, the number of boards are only produced limited numbers and the developers do not have his/her own board to build on it to check before the commits, many boards of engineering samples cost more than 500,000,000USD.

We have to adopt world if we do not have the alternative and we can not fix the world.
The cross-complier are mostly developed after the native compiler are working.

I feel it is OK to add cross-compiling feature after merging the new configure script.
Adding the make rpm later too.

**Comment by janstary** [comment]
I don't think the cross-compiling arguments above apply to our case.

On my [Beagle Bone Black](https://beagleboard.org/black) running OpenBSD 6.3/armv7,
`./configure ; make` takes about 12 seconds (with this new system).
That's hardly "so slow" we would need to cross-compile.

The arm* boards are pretty well supported nowadays,
at least they run an OS and a compiler (mostly various linuxes and BSDs).

I don't think cross-compilation is worth the added complexity.
If someone has an embedded board where he needs to run rtptools
but cannot compile natively, I will be glad to hear about his cool application and add it.

I agree that both this and the rpm (and possibly deb) packages
can be done _after_ this is merged in.

So, is there anything stopping us from merging?

**Comment by janstary** [comment]
BTW, the old `aclocal ; automake; autoconf; ./configure; make clean all` takes about 76s
(not counting the time to build autoconf, automake, pkg-config, help2man and xz),
out of which the actual `make clean all` is about 19s; that's about 400% overhead.

With this new system, `./configure ; make clean all` is about 12s,
out of which the actual `make clean all` is about 11s; that's about 10% overhead.

(Note how the new `./configure && make clean all` is faster, including ./configure,
than `make clean all` itself in the old system. Same `make(1)`, same `cc(1)`,
but a different `Makefile`.)

**Comment by mcd500** [comment]
> I don't think the cross-compiling arguments above apply to our case.

How about the other reason I mentioned?
I see only about comparing the compile speed in your comment against the Beagle Bone Black which is super fast already but how about other reasons that huge number of engineers are cross-compiling everyday.

Please try to build natively on any of the HP/Lexmark printers.
(Most of the time they even do not have compiler in the rootfs even during the development)

Software engineering is normally a team work.
Instead of making impression to others that anything you have never experienced is ignorable, 
it is always good to show that you are trying to support other persons perspective.

**Comment by janstary** [comment]
We already agreed to merge this. Can you please merge this?

If you need to run rtptools on a Lexmark printer,
or some othe platform that needs cross-compiling,
it's easy enough to add after this is in.

**Comment by mcd500** [comment]
Ok :)
On this thread, all what matter is the configure script.

> If you need to run rtptools on a Lexmark printer,
> or some othe platform that needs cross-compiling,

Lets not go on this road any further.

**Pull request merged**